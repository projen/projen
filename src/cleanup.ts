import * as fs from "fs";
import * as path from "path";
import * as glob from "fast-glob";
import { PROJEN_DIR, PROJEN_MARKER } from "./common";
import * as logging from "./logging";

export const FILE_MANIFEST = `${PROJEN_DIR}/files.json`;

export function cleanup(dir: string, newFiles: string[], exclude: string[]) {
  try {
    const manifestFiles = getFilesFromManifest(dir);
    if (manifestFiles.length > 0) {
      // Use `FILE_MANIFEST` to remove files that are no longer managed by projen
      removeFiles(findOrphanedFiles(dir, manifestFiles, newFiles));
    } else {
      // Remove all files managed by projen with legacy logic
      removeFiles(findGeneratedFiles(dir, exclude));
    }
  } catch (e: any) {
    logging.warn(`warning: failed to clean up generated files: ${e.stack}`);
  }
}

function removeFiles(files: string[]) {
  for (const file of files) {
    fs.rmSync(file, { force: true, recursive: true });
  }
}

function findGeneratedFiles(dir: string, exclude: string[]) {
  const ignore = [
    ...readGitIgnore(dir),
    "node_modules/**",
    ...exclude,
    ".git/**",
  ];

  const files = glob.sync("**", {
    ignore,
    cwd: dir,
    dot: true,
    onlyFiles: true,
    followSymbolicLinks: false,
    absolute: true,
  });

  const generated = new Array<string>();

  for (const file of files) {
    const contents = fs.readFileSync(file, "utf-8");

    if (contents.includes(PROJEN_MARKER)) {
      generated.push(file);
    }
  }

  return generated;
}

function findOrphanedFiles(
  dir: string,
  oldFiles: string[],
  newFiles: string[]
) {
  return oldFiles
    .filter((old) => !newFiles.includes(old))
    .map((f: string) => path.resolve(dir, f));
}

function getFilesFromManifest(dir: string): string[] {
  try {
    const fileManifestPath = path.resolve(dir, FILE_MANIFEST);
    if (fs.existsSync(fileManifestPath)) {
      const fileManifest = JSON.parse(
        fs.readFileSync(fileManifestPath, "utf-8")
      );
      if (fileManifest.files) {
        return fileManifest.files;
      }
    }
  } catch (e: any) {
    logging.warn(
      `warning: unable to get files to clean from file manifest: ${e.stack}`
    );
  }

  return [];
}

function readGitIgnore(dir: string) {
  const filepath = path.join(dir, ".gitignore");
  if (!fs.existsSync(filepath)) {
    return [];
  }

  return fs
    .readFileSync(filepath, "utf-8")
    .split("\n")
    .filter((x) => x?.trim() !== "")
    .filter((x) => !x.startsWith("#") && !x.startsWith("!"))
    .map((x) => x.replace(/^\//, "")) // remove "/" prefix
    .map((x) => `${x}\n${x}/**`)
    .join("\n")
    .split("\n");
}
