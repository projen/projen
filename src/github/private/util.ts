export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

export function context(value: string) {
  return `\${{ ${value} }}`;
}

// Checks if part of the file path is hidden
export function isHiddenPath(path: string) {
  return /(^|\/)\.[^\/\.]/g.test(path);
}

// Helper to assert a path is not hidden
export function ensureNotHiddenPath(value: string, name: string) {
  if (isHiddenPath(value)) {
    throw Error(`${name} cannot be a hidden path, got: ${value}`);
  }
}
