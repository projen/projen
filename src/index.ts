export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
