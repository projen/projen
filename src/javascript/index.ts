export * from "./bundler";
export * from "./eslint";
export * from "./jest";
export * from "./license-checker";
export * from "./node-package";
export * from "./node-project";
export * from "./npm-config";
export * from "./prettier";
export * from "./projenrc";
export * from "./typescript-config";
export * from "./upgrade-dependencies";
export * from "./yarnrc";
