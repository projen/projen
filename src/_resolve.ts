import { types } from "util";
import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // Special resolution for few JavaScript built-in types
  // that by default would be stringified as empty objects ('{}')
  // as they are missing a `toJSON` implementation.
  switch (true) {
    case types.isRegExp(value):
      if (value.flags) {
        throw new Error(
          "RegExp with flags should be explicitly converted to a string"
        );
      }
      return value.source;

    case types.isSet(value):
      return resolve(Array.from(value), options);

    case types.isMap(value):
      return resolve(Object.fromEntries(value), options);

    case types.isBoxedPrimitive(value):
      return resolve(value.valueOf(), options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
