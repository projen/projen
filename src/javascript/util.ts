import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve, posix } from "path";
import * as semver from "semver";
import { NodePackage } from "./node-package";
import { Project } from "../project";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(posix.sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return posix.join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given dependency version is installed
 *
 * This can be used to test for the presence of certain versions of devDependencies,
 * and do something dependency-specific in certain Components. For example, test for
 * a version of Jest and generate different configs based on the Jest version.
 *
 * NOTE: The implementation of this function currently is currently
 * approximate: to do it correctly, we would need a separate implementation
 * for every package manager, to query its installed version (either that, or we
 * would code to query `package-lock.json`, `yarn.lock`, etc...).
 *
 * Instead, we will look at `package.json`, and assume that the versions
 * picked by the package manager match ~that. This will work well enough for
 * major version checks, but may fail for point versions.
 *
 * What we SHOULD do is: `actualVersion ∈ checkRange`.
 *
 * What we do instead is a slightly more sophisticated version of
 * `requestedRange ∩ checkRange != ∅`. This will always give a correct result if
 * `requestedRange ⊆ checkRange`, but may give false positives when `checkRange
 * ⊆ requestedRange`.
 *
 * May return `undefined` if the question cannot be answered. These include the
 * following cases:
 *
 *   - The dependency is requested via local file dependencies (`file://...`)
 *   - The dependency uses an other type of URL, such as a GitHub URL
 *   - The dependency is not found in the `package.json`, such as when
 *     bootstrapping from an external projen package, and the `package.json`
 *     file only has that one external package as a dependency
 *
 * Otherwise it will return `true` if the installed version is probably in the
 * requested range, and `false` if it is probably not.
 *
 * This API may eventually be added to the public projen API, but only after
 * we implement exact version checking.
 *
 * @param dependencyName The name of the dependency
 * @param checkRange A particular version, or range of versions.
 */
export function hasDependencyVersion(
  project: Project,
  dependencyName: string,
  checkRange: string
): boolean | undefined {
  const file = NodePackage.of(project)?.file;
  if (!file) {
    return undefined;
  }

  if (!existsSync(file.absolutePath)) {
    return undefined;
  }

  const pj = JSON.parse(readFileSync(file.absolutePath, "utf-8"));

  // Technicaly, we should be intersecting all ranges to come up with the most narrow dependency
  // range, but `semver` doesn't allow doing that and we don't want to add a dependency on `semver-intersect`.
  //
  // Let's take the first dependency declaration we find, and assume that people
  // set up their `package.json` correctly.
  let requestedRange: string | undefined;
  for (const key of ["dependencies", "devDependencies", "peerDependencies"]) {
    const deps = pj[key] ?? {};
    let requestedVersion = deps[dependencyName];
    if (requestedVersion) {
      // If this is not a valid range, it could be 'file:dep.tgz', or a GitHub URL. No way to know what
      // version we're getting, bail out.
      if (!semver.validRange(requestedVersion)) {
        return undefined;
      }
      requestedRange = requestedVersion;
      break;
    }
  }

  // If the dependency is not found in the `package.json`, we can't answer the question (yet).
  // This means that the dependency hasn't been added yet, which means we know (upstream) what we're going to request,
  // or we're going to ask for '*' and we'll get the latest version.
  if (!requestedRange) {
    return undefined;
  }

  return installedVersionProbablyMatches(requestedRange, checkRange);
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
