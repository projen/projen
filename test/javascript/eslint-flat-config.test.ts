import { EslintFlatConfig, NodeProject } from "../../src/javascript";
import { synthSnapshot } from "../util";

describe("snapshot", () => {
  test("enable prettier", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new EslintFlatConfig(project, {
      enablePatterns: ["**/*.ts", "**/*.tsx"],
      prettier: true,
    });
    eslint.synthesize();

    // THEN
    expect(synthSnapshot(project)["eslint.config.mjs"]).toMatchSnapshot();
  });

  test("disable prettier", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new EslintFlatConfig(project, {
      enablePatterns: ["**/*.ts", "**/*.tsx"],
    });
    eslint.synthesize();

    // THEN
    expect(synthSnapshot(project)["eslint.config.mjs"]).toMatchSnapshot();
  });

  test("override setting", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new EslintFlatConfig(project, {
      enablePatterns: ["**/*.ts", "**/*.tsx"],
      prettier: true,
    });
    eslint.addOverrides({
      enablePatterns: ["projenrc.ts"],
      rules: {
        "@typescript-eslint/no-require-imports": "off",
        "import/no-extraneous-dependencies": "off",
      },
      parser: {
        importPath: "typescript-eslint",
        moduleName: "tsEslint",
        parserReference: "tsEslint.parser",
      },
    });
    eslint.synthesize();

    // THEN
    expect(synthSnapshot(project)["eslint.config.mjs"]).toMatchSnapshot();
  });
});

describe("eslint settings", () => {
  test("alias", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new EslintFlatConfig(project, {
      enablePatterns: ["**/*.ts", "**/*.tsx"],
      aliasMap: {
        "@src": "./src",
        "@foo": "./src/foo",
      },
      aliasExtensions: [".ts", ".tsx"],
    });
    eslint.synthesize();

    // THEN
    const pattern =
      /settings:\s*{[\s\S]*?"import\/resolver":\s*{([\s\S]*?)}[\s\S]*?}/;
    const settingsMatch = eslint.config.match(pattern);
    expect(settingsMatch).not.toBeNull();
    expect(settingsMatch![0]).toContain(
      '"alias":{"map":[["@src","./src"],["@foo","./src/foo"]],"extensions":[".ts",".tsx"]}'
    );
  });

  test("tsAlwaysTryTypes", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new EslintFlatConfig(project, {
      enablePatterns: ["**/*.ts", "**/*.tsx"],
      tsAlwaysTryTypes: true,
    });
    eslint.synthesize();

    // THEN
    const pattern =
      /settings:\s*{[\s\S]*?"import\/resolver":\s*{(\s*typescript:\s*{[\s\S]*?})\s*}[\s\S]*?}/;
    const settingsMatch = eslint.config.match(pattern);
    expect(settingsMatch).not.toBeNull();
    expect(settingsMatch![0]).toContain("alwaysTryTypes: true");
  });

  test("devDirs", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new EslintFlatConfig(project, {
      enablePatterns: ["**/*.ts", "**/*.tsx"],
      devDirs: ["src/foo", "src/bar"],
    });
    eslint.synthesize();

    // THEN
    expect(eslint.enablePatterns.includes("src/foo")).toStrictEqual(true);
    expect(eslint.enablePatterns.includes("src/bar")).toStrictEqual(true);
    expect(eslint.rules["import/no-extraneous-dependencies"]).toEqual([
      "error",
      {
        devDependencies: ["**/src/foo/**", "**/src/bar/**"],
        optionalDependencies: false,
        peerDependencies: true,
      },
    ]);
  });

  test("can add enable patterns", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new EslintFlatConfig(project, {
      enablePatterns: ["**/*.ts", "**/*.tsx"],
    });
    eslint.addEnablePatterns("src/foo", "src/bar");

    // THEN
    expect(eslint.enablePatterns.includes("src/foo")).toStrictEqual(true);
    expect(eslint.enablePatterns.includes("src/bar")).toStrictEqual(true);
  });

  test("can add ignore patterns", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new EslintFlatConfig(project, {
      enablePatterns: ["**/*.ts", "**/*.tsx"],
    });
    eslint.addIgnorePatterns("src/foo", "src/bar");

    // THEN
    expect(eslint.ignorePatterns.includes("src/foo")).toStrictEqual(true);
    expect(eslint.ignorePatterns.includes("src/bar")).toStrictEqual(true);
  });

  test("can add rules", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new EslintFlatConfig(project, {
      enablePatterns: ["**/*.ts", "**/*.tsx"],
    });
    eslint.addRules({
      "@typescript-eslint/no-require-imports": "off",
      "import/no-extraneous-dependencies": "off",
    });

    // THEN
    expect(eslint.rules["@typescript-eslint/no-require-imports"]).toEqual(
      "off"
    );
    expect(eslint.rules["import/no-extraneous-dependencies"]).toEqual("off");
  });

  test("can add plugins", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new EslintFlatConfig(project, {
      enablePatterns: ["**/*.ts", "**/*.tsx"],
    });
    eslint.addPlugins({
      importPath: "eslint-plugin-regexp",
      moduleName: "regexpPlugin",
      pluginAlias: "regexp",
    });
    eslint.synthesize();

    // THEN
    const pattern = /plugins:\s*{[\s\S]*?}/;
    const pluginMatch = eslint.config.match(pattern);
    expect(pluginMatch).not.toBeNull();
    expect(pluginMatch![0]).toContain('"regexp": regexpPlugin');
  });

  test("can add extends", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new EslintFlatConfig(project, {
      enablePatterns: ["**/*.ts", "**/*.tsx"],
    });
    eslint.addExtends({
      importPath: "typescript-eslint",
      moduleName: "tseslint",
      configReference: "tseslint.config.recommended",
      shouldSpreadConfig: true,
    });
    eslint.synthesize();

    // THEN
    const pattern =
      /export\s+default\s+\[[\s\S]*?,\s*(\.{3}tseslint\.config\.recommended)[\s\S]*?\]/;
    expect(pattern.test(eslint.config)).toStrictEqual(true);
  });
});
