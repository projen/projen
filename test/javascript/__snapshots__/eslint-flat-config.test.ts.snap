// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot disable prettier 1`] = `
"
import globals from "globals";
import tseslint from "typescript-eslint"
import importPlugin from "eslint-plugin-import"
import stylistic from "@stylistic/eslint-plugin"
import eslint from "@eslint/js"    

/** @type {import('eslint').Linter.Config[]} */
export default [
  eslint.configs.recommended,
  importPlugin.flatConfigs.typescript,
  {
    files: ["**/*.ts", "**/*.tsx"],
    languageOptions: { 
      globals: {
        ...globals.node,
        ...globals.jest,
      },
      parser: tseslint.parser,
      parserOptions: {
        ecmaVersion: 2018,
        sourceType: "module",
        project: "./tsconfig.json",
      },
    },
    settings: {
      "import/parsers": {
        "@typescript-eslint/parser": [".ts", ".tsx"],
      },
      "import/resolver": {
        typescript: {
          project: "./tsconfig.json",
          alwaysTryTypes: true,
        }
      },
    },
    plugins: {
      "@typescript-eslint": tseslint,
      "import": importPlugin,
      "@stylistic": stylistic
    },
    rules: {
      "@typescript-eslint/no-require-imports": "error",
      "import/no-extraneous-dependencies": ["error",{"devDependencies":[],"optionalDependencies":false,"peerDependencies":true}],
      "import/no-unresolved": "error",
      "import/order": ["warn",{"groups":["builtin","external"],"alphabetize":{"order":"asc","caseInsensitive":true}}],
      "import/no-duplicates": "error",
      "no-shadow": "off",
      "@typescript-eslint/no-shadow": "error",
      "@stylistic/key-spacing": "error",
      "@stylistic/no-multiple-empty-lines": "error",
      "@typescript-eslint/no-floating-promises": "error",
      "no-return-await": "off",
      "@typescript-eslint/return-await": "error",
      "@stylistic/no-trailing-spaces": "error",
      "dot-notation": "error",
      "no-bitwise": "error",
      "@typescript-eslint/member-ordering": ["error",{"default":["public-static-field","public-static-method","protected-static-field","protected-static-method","private-static-field","private-static-method","field","constructor","method"]}],
      "indent": "off",
      "@stylistic/indent": ["error",2],
      "@stylistic/quotes": ["error","single",{"avoidEscape":true}],
      "@stylistic/comma-dangle": ["error","always-multiline"],
      "@stylistic/comma-spacing": ["error",{"before":false,"after":true}],
      "@stylistic/no-multi-spaces": ["error",{"ignoreEOLComments":false}],
      "@stylistic/array-bracket-spacing": ["error","never"],
      "@stylistic/array-bracket-newline": ["error","consistent"],
      "@stylistic/object-curly-spacing": ["error","always"],
      "@stylistic/object-curly-newline": ["error",{"multiline":true,"consistent":true}],
      "@stylistic/object-property-newline": ["error",{"allowAllPropertiesOnSameLine":true}],
      "@stylistic/keyword-spacing": "error",
      "@stylistic/brace-style": ["error","1tbs",{"allowSingleLine":true}],
      "@stylistic/space-before-blocks": "error",
      "curly": ["error","multi-line","consistent"],
      "@stylistic/member-delimiter-style": "error",
      "@stylistic/semi": ["error","always"],
      "@stylistic/max-len": ["error",{"code":150,"ignoreUrls":true,"ignoreStrings":true,"ignoreTemplateLiterals":true,"ignoreComments":true,"ignoreRegExpLiterals":true}],
      "@stylistic/quote-props": ["error","consistent-as-needed"]
    }
  },
  {
    ignores: ["*.js", "*.d.ts", "node_modules/", "*.generated.ts", "coverage"]
  }
];"
`;

exports[`snapshot enable prettier 1`] = `
"
import globals from "globals";
import tseslint from "typescript-eslint"
import importPlugin from "eslint-plugin-import"
import eslint from "@eslint/js"
import prettierConfig from "eslint-config-prettier"    

/** @type {import('eslint').Linter.Config[]} */
export default [
  eslint.configs.recommended,
  importPlugin.flatConfigs.typescript,
  prettierConfig,
  {
    files: ["**/*.ts", "**/*.tsx"],
    languageOptions: { 
      globals: {
        ...globals.node,
        ...globals.jest,
      },
      parser: tseslint.parser,
      parserOptions: {
        ecmaVersion: 2018,
        sourceType: "module",
        project: "./tsconfig.json",
      },
    },
    settings: {
      "import/parsers": {
        "@typescript-eslint/parser": [".ts", ".tsx"],
      },
      "import/resolver": {
        typescript: {
          project: "./tsconfig.json",
          alwaysTryTypes: true,
        }
      },
    },
    plugins: {
      "@typescript-eslint": tseslint,
      "import": importPlugin
    },
    rules: {
      "@typescript-eslint/no-require-imports": "error",
      "import/no-extraneous-dependencies": ["error",{"devDependencies":[],"optionalDependencies":false,"peerDependencies":true}],
      "import/no-unresolved": "error",
      "import/order": ["warn",{"groups":["builtin","external"],"alphabetize":{"order":"asc","caseInsensitive":true}}],
      "import/no-duplicates": "error",
      "no-shadow": "off",
      "@typescript-eslint/no-shadow": "error",
      "@stylistic/key-spacing": "error",
      "@stylistic/no-multiple-empty-lines": "error",
      "@typescript-eslint/no-floating-promises": "error",
      "no-return-await": "off",
      "@typescript-eslint/return-await": "error",
      "@stylistic/no-trailing-spaces": "error",
      "dot-notation": "error",
      "no-bitwise": "error",
      "@typescript-eslint/member-ordering": ["error",{"default":["public-static-field","public-static-method","protected-static-field","protected-static-method","private-static-field","private-static-method","field","constructor","method"]}]
    }
  },
  {
    ignores: ["*.js", "*.d.ts", "node_modules/", "*.generated.ts", "coverage"]
  }
];"
`;

exports[`snapshot override setting 1`] = `
"
import globals from "globals";
import tseslint from "typescript-eslint"
import importPlugin from "eslint-plugin-import"
import eslint from "@eslint/js"
import prettierConfig from "eslint-config-prettier"    

/** @type {import('eslint').Linter.Config[]} */
export default [
  eslint.configs.recommended,
  importPlugin.flatConfigs.typescript,
  prettierConfig,
  {
    files: ["**/*.ts", "**/*.tsx"],
    languageOptions: { 
      globals: {
        ...globals.node,
        ...globals.jest,
      },
      parser: tseslint.parser,
      parserOptions: {
        ecmaVersion: 2018,
        sourceType: "module",
        project: "./tsconfig.json",
      },
    },
    settings: {
      "import/parsers": {
        "@typescript-eslint/parser": [".ts", ".tsx"],
      },
      "import/resolver": {
        typescript: {
          project: "./tsconfig.json",
          alwaysTryTypes: true,
        }
      },
    },
    plugins: {
      "@typescript-eslint": tseslint,
      "import": importPlugin
    },
    rules: {
      "@typescript-eslint/no-require-imports": "error",
      "import/no-extraneous-dependencies": ["error",{"devDependencies":[],"optionalDependencies":false,"peerDependencies":true}],
      "import/no-unresolved": "error",
      "import/order": ["warn",{"groups":["builtin","external"],"alphabetize":{"order":"asc","caseInsensitive":true}}],
      "import/no-duplicates": "error",
      "no-shadow": "off",
      "@typescript-eslint/no-shadow": "error",
      "@stylistic/key-spacing": "error",
      "@stylistic/no-multiple-empty-lines": "error",
      "@typescript-eslint/no-floating-promises": "error",
      "no-return-await": "off",
      "@typescript-eslint/return-await": "error",
      "@stylistic/no-trailing-spaces": "error",
      "dot-notation": "error",
      "no-bitwise": "error",
      "@typescript-eslint/member-ordering": ["error",{"default":["public-static-field","public-static-method","protected-static-field","protected-static-method","private-static-field","private-static-method","field","constructor","method"]}]
    }
  },
  {
    languageOptions: {
      parser: tseslint.parser
    },
    files: ["projenrc.ts"],
    rules: {
      "@typescript-eslint/no-require-imports": "off",
      "import/no-extraneous-dependencies": "off"
    }
  }
  {
    ignores: ["*.js", "*.d.ts", "node_modules/", "*.generated.ts", "coverage"]
  }
];"
`;
